Automattic/vip-go-mu-plugins
# Prometheus: re-add the dedicated endpoint with better locking mechanism, introduce Error stats collector

## Description

This is a significant re-work of #3989, it contains the following changes:

1. Collector interface is amended to include a new `process_metrics` method. All collectors had that implemented.
- The purpose of the new method is to do any potentially heavy computations off the web request path and store it into temporary storage, so that `collect_metrics` can grab that data without having to waste time. See Post Stats Collector for example implementation.
- A new cron event is added `vip_prom_process_metrics`. It's get scheduled at around ~10 min interval (as a single event). This kicks off the aforementioned `process_metrics` for each collector.
- On the web request's shutdown on authenticated requests (might change in the future depending on initial rollout results) every 90 seconds we'll collect the metrics and put it into APCu for the scraper to collect.
2. Error Stats collector is introduced, to report the number of runtime error per network site and error type
- To achieve that our current error handler adds a new action `php_error_handler`. `do_action` gets called if it's defined, Error Stats Collector hooks onto that to increment the relevant counter.


## Checklist

Please make sure the items below have been covered before requesting a review:


## Steps to Test

1. Check out the PR
4. For ease of testing, comment out the current_user_can. 
5. Add an error_log inside if `( wp_cache_add( 'vip_prometheus_last_collection_run', time(), 'vip_prometheus', 5 * MINUTE_IN_SECONDS ) ) {`
6. Comment out the auth check
7. Kick off something like `ab -c 10 -n 300 http://site.vipdev.lndo.site`
8. Make sure there's just one message logged (i.e. the locking works)
9. On the v2.27.0 of VIP-CLI, visit `http://site.vipdev.lndo.site/.vip-prom-metrics` and verify it's rendered correctly.
10. Add some test code like `trigger_error( 'My test warning', E_USER_WARNING );`, verify that `error_count` is present in the prev step.
11. Disable the Search collector (by nerfing the method) and verify Search is working in all contexts and there are no fatals.
